globals.go

```diff
@@ -0,0 +1,208 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terramate
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	tflang "github.com/hashicorp/terraform/lang"
+	"github.com/madlambda/spells/errutil"
+	"github.com/mineiros-io/terramate/config"
+	"github.com/mineiros-io/terramate/hcl"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Globals represents a globals block.
+type Globals struct {
+	attributes map[string]cty.Value
+}
+
+const ErrGlobalRedefined errutil.Error = "global redefined"
+
+// LoadStackGlobals loads from the file system all globals defined for
+// a given stack. It will navigate the file system from the stack dir until
+// it reaches rootdir, loading globals and merging them appropriately.
+//
+// More specific globals (closer or at the stack) have precedence over
+// less specific globals (closer or at the root dir).
+//
+// Metadata for the stack is used on the evaluation of globals, defined on stackmeta.
+// The rootdir MUST be an absolute path.
+func LoadStackGlobals(rootdir string, meta StackMetadata) (*Globals, error) {
+	if !filepath.IsAbs(rootdir) {
+		return nil, fmt.Errorf("%q is not absolute path", rootdir)
+	}
+
+	unEvalGlobals, err := loadStackGlobals(rootdir, meta.Path)
+	if err != nil {
+		return nil, err
+	}
+	return unEvalGlobals.eval(meta)
+}
+
+// Attributes returns all the global attributes, the key in the map
+// is the attribute name with its corresponding value mapped
+func (g *Globals) Attributes() map[string]cty.Value {
+	return g.attributes
+}
+
+func (g *Globals) add(name string, val cty.Value) {
+	g.attributes[name] = val
+}
+
+type rawGlobals struct {
+	expressions map[string]hclsyntax.Expression
+}
+
+func (r *rawGlobals) merge(other *rawGlobals) {
+	for k, v := range other.expressions {
+		if !r.has(k) {
+			r.add(k, v)
+		}
+	}
+}
+
+func (r *rawGlobals) add(name string, expr hclsyntax.Expression) {
+	r.expressions[name] = expr
+}
+
+func (r *rawGlobals) has(name string) bool {
+	_, ok := r.expressions[name]
+	return ok
+}
+
+func (r *rawGlobals) eval(meta StackMetadata) (*Globals, error) {
+	// TODO(katcipis): add BaseDir on Scope.
+	tfscope := &tflang.Scope{}
+	evalctx, err := newHCLEvalContext(meta, tfscope)
+	if err != nil {
+		return nil, err
+	}
+
+	// error messages improve if globals is empty instead of undefined
+	empty, err := hclMapToCty(nil)
+	if err != nil {
+		return nil, fmt.Errorf("initializing empty globals: %v", err)
+	}
+
+	evalctx.Variables["global"] = empty
+
+	var errs []error
+	globals := newGlobals()
+	pendingExprs := r.expressions
+
+	for len(pendingExprs) > 0 {
+		evaluated := 0
+
+		for name, expr := range pendingExprs {
+			val, err := expr.Value(evalctx)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			globals.add(name, val)
+			evaluated += 1
+			delete(pendingExprs, name)
+		}
+
+		if evaluated == 0 {
```

Comment:
if we call this var `amountEvaluated` or something like that implying this *if* is about `if noneEvaluated` then break. Just sharing because at the beginning I thought you were missing the case `evaluate > 0` here. But feel free to ignore, it just took more time to understand. Nice implementation though =)

Reply:
will rename, makes sense, thanks man :100: